core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -247,6 +247,34 @@ final class KafkaMetadataLog private (
     FileRawSnapshotWriter.create(log.dir.toPath, snapshotId, Optional.of(this))
   }
 
+  override def createSnapshotFromEndOffset(endOffset: Long): RawSnapshotWriter = {
+    val highWatermarkOffset = highWatermark.offset
+    if (endOffset > highWatermarkOffset) {
+      throw new IllegalArgumentException(
+        s"Cannot create a snapshot for an end offset ($endOffset) greater than the high-watermark ($highWatermarkOffset)"
+      )
+    }
+
+    if (endOffset < startOffset) {
+      throw new IllegalArgumentException(
+        s"Cannot create a snapshot for an end offset ($endOffset) less or equal to the log start offset ($startOffset)"
```

Comment:
nit: the code is `endOffset < startOffset` but the log is `less or equal to`.

Reply:
Fixed as part of the merge commit.